app/models/show_product.rb

```diff
@@ -0,0 +1,125 @@
+class ShowProduct
+  attr_reader :acting_user, :controller, :error, :login_required, :redirect, :session_user
+
+  def initialize(controller, product, acting_user, session_user, acting_as)
```

Comment:
Since you're passing a reference to the controller in, you could probably `delegate :acting_user, :session_user, :acting_as?, to: :controller`, and then you wouldn't need as many parameters and instance variables.